src/adera/ShipResources.cpp

```diff
@@ -0,0 +1,119 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "ShipResources.h"
+#include "osp/Active/ActiveScene.h"
+
+using namespace osp;
+using namespace osp::active;
+using namespace adera::active::machines;
+
+/* MachineContainer */
+
+void MachineContainer::propagate_output(WireOutput* output)
+{
+
+}
+
+WireInput* MachineContainer::request_input(WireInPort port)
+{
+    return nullptr;
+}
+
+WireOutput* MachineContainer::request_output(WireOutPort port)
+{
+    return nullptr;
+}
+
+std::vector<WireInput*> MachineContainer::existing_inputs()
+{
+    return m_inputs;
+}
+
+std::vector<WireOutput*> MachineContainer::existing_outputs()
```

Comment:
These should be returning by const&.

Otherwise, you're forcing the caller to always make a copy, when maybe they just want to reference what you got.